### ISSUE TITLE
New Theme Update Issue

### ISSUE THREAD

**Issue by iedmrc** [bitnami/bitnami-docker-opencart#51]
Hi,
I use bitnami-docker-opencart image smoothly but with a problem. This is a part from my docker-compose.yml:

```
opencart:
    image: 'bitnami/opencart:latest'
    environment:
      - ALLOW_EMPTY_PASSWORD=yes
      - OPENCART_USERNAME=blabla
      - OPENCART_PASSWORD=blabla
      - OPENCART_EMAIL=blabla
      - OPENCART_HOST=blabla
      - MARIADB_PASSWORD=blabla
    ports:
      - '80:80'
      - '443:443'
    volumes:
      - '../web_app:/bitnami'
    depends_on:
      - mariadb
```

I uploaded my new theme to web_app local directory and i can see it is mounted to opencart container very well. When I enter into container with exec command, i can see files under `/bitnami/opencart` directory. But bitnami runs apache from /opt/bitnami/opencart and it doesn't sync with `/bitnami/opencart` directory. So i cannot use my new theme with bitnami-opencart- image.

Why this image works with this way? What should be the best practice to solve this problem?

Thank you for your helps.

**Comment by juan131** [comment]
Hello @dmrc1143 

Could you please follow this workaround

 - Create a local directory to store the OpenCart Data.

```
mkdir opencart_data
```

 - Modify the `docker-compose.yml` volumed section:
 
```
volumes:
- './opencart_data/:/bitnami/opencart'
```

 - The application would be installed and you'll have the OpenCart data on your local directory `opencart_data`. You can check the content of the directory have all the Open Cart data:

```
$ tree opencart_data
opencart_data/
├── admin
│   ├── config.php
│   └── view
│       └── stylesheet
│           └── bootstrap.css
├── config.php
├── image
│   ├── cache
│   │   └── index.html
│   ├── catalog
│   │   ├── cart.png
│   │   ├── demo
│   │   │   ├── apple_cinema_30.jpg

...
```

You can include your theme there.

**Comment by dmrc1143** [comment]
okey, it seems i can store data here. But if i want to mount whole opencart files from my local, what would be the correct way to do ? How can i mount whole opencart files from my local to bitnami_opencart container?

**Comment by juan131** [comment]
You should only mount the "data" files. The rest of the files should not be moved to the container since they will be automatically retrieved. 

You can also move your data using the Import/Export extensions OpenCart offers you. I mean... you can export your products, etc. using a CSV file and import them on the new OpenCart container.

**Comment by rfvallina** [comment]
Hello,
I'm using image bitnami/opencart:3.0.2-0.
I have the same problem. Why /bitnami is not sync'ed with /opt/bitnami?
Is it mandatory to point a host path to sync data?

**Comment by juan131** [comment]
Hello @rfvallina 

> Why /bitnami is not sync'ed with /opt/bitnami?

**/bitnami** is the directory used to persist data. Therefore, any other file that's not considered configuration or data won't be stored there (such as the php code). As you can check, there are symbolic links from the corresponding directory at **/opt/bitnami/** to **/bitnami**.

That's important because you can upgrade your app's code doing a simple restart of the container. During the second initialisation of your container everything will be re-generated (using the new version of the code), except the data persisted at **/bitnami**.

> Is it mandatory to point a host path to sync data?

You can use a Docker Volume to save your data too (more info: https://docs.docker.com/storage/volumes/)

**Comment by iedmrc** [comment]
Hi @rfvallina ,
I just fixed this issue with overriding app-entrypoint.sh (with a new Dockerfile) with this:

`#!/bin/bash -e

. /opt/bitnami/base/functions
. /opt/bitnami/base/helpers

print_welcome_page
check_for_updates &

if [[ "$1" == "nami" && "$2" == "start" ]] || [[ "$1" == "/init.sh" ]]; then
  nami_initialize apache php opencart
  info "Starting opencart... "
fi

rm -rf /opt/bitnami/opencart
ln -s /bitnami/opencart/ /opt/bitnami/opencart

exec tini -- "$@"`

these lines do the trick:
`rm -rf /opt/bitnami/opencart
ln -s /bitnami/opencart/ /opt/bitnami/opencart`

**Comment by rfvallina** [comment]
Thanks for the reply @juan131 and @iedmrc.

The problem I have is the extensions I'm installing are not persistent. I guess this is because they add custom php files and they don't have a corresponding link to /opt/bitnami/opencart/... 

Then, what would be the best way to solve this and have all my extensions persisted?
Should I create a custom image which has bitnami/opencart as base image?

**Comment by iedmrc** [comment]
as I wrote, I used a ne dockerfile with these two lines:

```
FROM bitnami/opencart:latest
COPY ./app-entrypoint.sh /app-entrypoint.sh
```
and with this way overrided app-entrypoint.sh .
Mind that, I still have the same docker-compose file as in the [first message](https://github.com/bitnami/bitnami-docker-opencart/issues/51#issue-278013152)

With this workaround, you get all of your opencart files from your local to container.
That means, for example, if you remove a php file from your local (opencart dir), it will be removed from container inside. Or if you add one or more files , it will be ...

**Comment by rfvallina** [comment]
@iedmrc So you run first your dockerfile and after that you run docker-compose?

**Comment by iedmrc** [comment]
I just created a new image with this dockerfile. Please see the reference sheet:
https://docs.docker.com/compose/compose-file

**Comment by rfvallina** [comment]
Thanks! 
I created a Dockerfile copying the entry point as you stated and it's working fine.

Below is my Dockerfile:

`FROM bitnami/opencart:3.0.2-0

COPY ./app-entrypoint.sh /app-entrypoint.sh
RUN ["chmod", "+x", "/app-entrypoint.sh"]`

And I run compose like this:

`docker-compose up --build -d`

This way, it recreates the image from the Dockerfile

**Comment by danielbjornadal** [comment]
@iedmrc & @rfvallina 
I cant get this to work.

The original /opt/bitnami/opencart folder contains data not present in /bitnami/opencart
Bitnami should really fix this image by making it so that /opt/bitnami only contains static data.

**Comment by juan131** [comment]
Hi @danielbjornadal 

> The original /opt/bitnami/opencart folder contains data not present in /bitnami/opencart

What data do you refer to? Could you please detail what directories/files at **/opt/bitnami/opencart** you consider that should be persisted at **bitnami/opencart**?

**Comment by danielbjornadal** [comment]
> What data do you refer to? Could you please detail what directories/files at **/opt/bitnami/opencart** you consider that should be persisted at **bitnami/opencart**?

Hi @juan131 

When adding new themes, languages or extensions, these must be copied to
`/opt/bitnami/opencart/` and `/opt/bitnami/opencart/catalog/language/`

There is currently no easy way of persisting `/opt/bitnami/` on the host system.

As a workaround for us, we had to rebuild the image with all extensions etc.

**Comment by juan131** [comment]
Hi @danielbjornadal 

Thanks for the information. I'm going to create a task to revisit what data we're persisting on Opencart so we can improve the way you add you custom themes or extensions. We need to find a way to persist as much as data as possible without breaking the upgrade method.

**Comment by medoix** [comment]
Any updates on this? Is it possible to just link /opt/bitnami/opencart to the mounted volume? If OpenCart needs to update any of the files it still can..

**Comment by tompizmor** [comment]
Hi @medoix,

I am afraid we don't have any news with regard to this issue. 

The problem linking the whole installation directory (`/opt/bitnami/opencart`) to the mounted volume (`/bitnami/opencart`) is that the upgrades for Opencart will stop working as you are persisting the entire application. Sometimes it is preferred easy development experience rather than upgradability but as I said before we couldn't evaluate the options for Opencart yet.

**Comment by Lenny-ai** [comment]
Hi, 
I followed the suggestion of @danielbjornadal which worked out fine initially. However, I now experience the following issue during the upgrading phase of opencart initialization:

opencar INFO  Upgrading OpenCart...
opencart_1  | opencar DEBUG ==> Setting timeout to 120000 miliseconds
opencart_1  | Error executing 'postInstallation': Server returned status code 500 from http://127.0.0.1/install/index.php?route=upgrade/upgrade
opencart_1  | Server returned status code 500 from http://127.0.0.1/install/index.php?route=upgrade/upgrade

Has anyone come across this ever since?

**Comment by juan131** [comment]
Hi @Lenny-ai 

It's hard to debug what's going on unless you access the container and check the Apache logs. 

> I followed the suggestion of @danielbjornadal which worked out fine initially.

Could you please specify the exact modifications you did in the themes, extensions, etc. and how did you perform them?

**Comment by Lenny-ai** [comment]
Hi @juan131, 

Accessing th Apache logs
How is the opencart_1 container accessible? It is not completely build. It stops building the container with "excited with code 1" and a flashing cursor .  
I tried to access the container with: docker exec -it <containername> bash 
which returns an error: No such container

Modifications I performed:
1) Create opencart_data folder
2) Copy an existing opencart project into this folder
3) use the docker-compose.yml file to mount this folder as a volume with /bitnami in the container
4) use the app-entrypoint.sh file and add the lines:
rm -rf /opt/bitnami/opencart
ln -s /bitnami/opencart/ /opt/bitnami/opencart/
5) use the yml-file also to mount a volume for the mariadb_data 
6) docker-compose up --build
7) in the admin area fo opencart uninstall and install all themes and extensions
8) backup the DB of the existing opencart project and restore it in the dockerized opencart project

That was all fine and I could recreate apperance and functionallity in the dockerized project.

9) stopped the running docker instances with CTRL+C
10) trying to docker-compose up --build

And now running into the issue stated above

I tried so many things in the meantime including chekcing permissions and config.php settings without any success.

Also I tried to get more information from the upgrading phase with:
environment:
 - NAMI_DEBUG='--log-level=trace8'
However this doesn't provide any log-level info of the upgrading phase.

**Comment by juan131** [comment]
> By the way: Aren't the two lines in the Dockerfile exactly doing that ... echoing all errors and accesses to the terminal?

You're totally right. The Apache logs are being redirected to stderr/stdout.

> 2) Copy an existing opencart project into this folder
> ...
> 4) use the app-entrypoint.sh file and add the lines:
> rm -rf /opt/bitnami/opencart
> ln -s /bitnami/opencart/ /opt/bitnami/opencart

I don't think such a dramatic change it's going to be compatible with the way Nami works and how Opencart is initialised. Why don't you just copy the extensions/themes you want to change instead of removing the whole `/opt/bitnami/opencart` directory? Sth like the solution @danielbjornadal suggested:

```
FROM bitnami/opencart:3.0.2-0-debian-9

COPY themes/oc-light/ /opt/bitnami/opencart/
COPY extensions/bambora/ /opt/bitnami/opencart/
```

**Comment by Lenny-ai** [comment]
Hi @juan131 

thanks for the suggestion. However I modified PHP files on the existing opencart project I would now like to dockerize. That is why I proceed with the brute force approach. 

In the meantime I found the root cause for the problem. Opencart was missing a folder and files of a theme I installed manually in the admin section. After stopping docker and rebuiding with docker-compose up -- build Opencart was missing something in /opt/bitnami/opencart/admin/view/template/extension

I found out by accidentally looking into the opencart files in the mounted volume: opencart_data/opencart/system/stroage/logs/error.log >>> that pointed me to a php error and the missing folder and file.

The solution was:
In the file add-entrypoint.sh, inside the if loop: split the nami_initialize and include a copy command. It now looks like this:
`  
if [[ "$1" == "nami" && "$2" == "start" ]] || [[ "$1" == "httpd" ]]; then
  . /apache-init.sh
  . /opencart-init.sh

  nami_initialize apache php mysql-client

  cp -r /bitnami/opencart/admin/view/template/extension/<missing folder> /opt/bitnami/opencart/admin/view/template/extension

  nami_initialize  opencart
  info "Starting opencart... "
fi
`
Now it builds the conatiner smoothly again.

What would have saved me three days would be some code that echoes opencart error messages to the terminal. So far the dockerfile only echoes Apache errors. @juan131 could you please suggest a solution for echoing opencart error messages.

Thanks anyway ... at least I have learned something new and that is what counts!
Cheers.

**Comment by juan131** [comment]
HI @Lenny-ai 

I'm glad you were able to identify the issue. Thanks so much for sharing the solution since it'll help others user that are trying to customize their own Opencart image.

>  So far the dockerfile only echoes Apache errors. @juan131 could you please suggest a solution for echoing opencart error messages.

What log file are you interested on redirecting to stdout/stderr? You can add an instruction like the one below to your Dockerfile so the logs are redirected to `/stdout/stderr`:

```
RUN  ln -sf "/dev/stdout" /path/to/your/access_log
RUN  ln -sf "/dev/stderr" /path/to/your/error_log
```

**Comment by marcosbc** [comment]
Hi, we have just released [OpenCart `3.0.3-6-debian-10-r65`](https://github.com/bitnami/bitnami-docker-opencart#303-6-debian-10-r65) which includes major changes from previous releases. Now, source code for the initialization logic is published in GitHub in the [`rootfs` folder](https://github.com/bitnami/bitnami-docker-opencart/tree/master/3/debian-10/rootfs) and also introduces significant changes to the persistence approach.

In particular, now it is possible for you to specify custom list of files to persist via the environment variable [`OPENCART_DATA_TO_PERSIST`](https://github.com/bitnami/bitnami-docker-opencart/blob/master/3/debian-10/rootfs/opt/bitnami/scripts/opencart-env.sh#L62) in your docker-compose (or Docker env).
